### ISSUE TITLE
Some parts of PyTorch test suite don't work properly on non-default stream

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#21589]
In @fbhuba PR https://github.com/pytorch/pytorch/pull/21474 we switch the stream before running tests. But in this PR, a lot of tests failed and we had to disable the stream switching behavior for large portions of the test suite. There is probably something very important in PyTorch that is not stream safe, we need to figure it out.

Currently known reasons why you might be on wrong stream:

* Calling thrust functions (e.g., `thrust::tabulate`) without specifying a policy that puts things on the right stream
* Backwards runs on default stream

**Comment by smileeok** [comment]
The performance on non-default stream is much worse than the default stream, as we tested. Just an FYI.

**Comment by ngimel** [comment]
Do we have a CI link to look at failed tests? Also, something demonstrating worse performance on non-default streams?

**Comment by smileeok** [comment]
When we use multi-streams, we noticed a bunch of cuda memory allocation and free as well as cuda event destroys that do not exist when we just use the default stream.

**Comment by ngimel** [comment]
Do you have a small script to reproduce that?

**Comment by jeffreyksmithjr** [comment]
We should probably keep the scope of this ticket to correctness and not performance. @smileeok can you file a separate issue around the performance concerns. @ngimel I think we use the CI-based test failures from #21474 as the issue to be addressed in this PR.

**Comment by ezyang** [comment]
> Do we have a CI link to look at failed tests?

@ngimel I observed it's a lot of "the number is wrong" type situation. I sent in a PR https://github.com/pytorch/pytorch/pull/21595 which turns on all the non-working tests so that you can poke at the CI.

**Comment by ngimel** [comment]
Errors in EmbeddingBag tests are arguably incorrect tests. It is well known that backward always happens on the default stream, regardless of the stream used by forward (which is a big inconvenience, btw, but a subject for a separate conversation) (#7601). EmbeddingBag test runs backward on default stream (no choice), then starts comparing results on the non-blocking non-default stream when results are not yet ready https://github.com/pytorch/pytorch/blob/a93a1ccbb375d1629d28ea1c16067ae394869dcb/test/test_nn.py#L2407-L2419. Adding synchronization after .backward calls fixes it. 
This can also happen in real world situations, when e.g. grad norm or optimizer will start running before backward is ready if user is trying to use non-default stream for them, because the fact that backward runs on default stream is not obvious and not documented.

**Comment by ezyang** [comment]
Oooh, I had forgotten about that. We should probably just fix that, then, and a bunch more things might work.

@malvika2147 is working on autograd this summer, this might be a good thing to work on.

**Comment by ngimel** [comment]
Same thing is happening in common_nn tests https://github.com/pytorch/pytorch/blob/master/test/common_nn.py#L3240-L3241, so short term fix to make tests pass is to put syncs after backward where necessary, but long term backward should respect forward stream, and then those syncs won't be needed.

**Comment by ezyang** [comment]
I'm reducing priority because this has been broken for a while, and we don't have very many users demanding that non-standard streams work. If we end up with an important use case for non-standard streams bump this up.

**Comment by ezyang** [comment]
@ngimel tells me that getting CUDA graphs https://github.com/pytorch/pytorch/issues/15623 to work is blocked on this working correctly (because CUDA graphs must run on a non-default stream.) So when CUDA graphs get to the point where they are usable and we want to let people use them, we will need to fix this.

**Comment by ezyang** [comment]
Back in December, I remember we had a bit of an argument about what the CUDA stream we should put PyTorch kernels on by default (PyTorch's historical behavior was using the legacy default stream; Caffe2's historical behavior was using a freshly allocated stream). One point of matter we did agree on was that, regardless of what we decided to do in the end, we should run PyTorch's test suite on a non default stream to flush out bugs. I remember someone (was it you, @zheng-xq?) making a statement that it would be "very obvious" if a kernel failed to use the correct stream.

Experience with #21785 suggests that code that is not stream safe leads to nondeterministic bugs which only fail occasionally, and only fail when you run the test suite in its entirety (as opposed to running the test by itself). This is BAD, especially because @ngimel has pointed out there are a number of places where we are doing obviously the wrong thing.

Thus, my suggestion, is that we should NOT enable this testing by default, not until we make an honest effort to fix the backwards sync problem.  #21793 disables this testing.

cc @dzhulgakov @ilia-cher

**Comment by gchanan** [comment]
Ya, we basically need a point person to handle fixing this and to follow test failures until they are fixed.

**Comment by SsnL** [comment]
So, out of curiosity, other than autograd, thrust, and rng functions, what is known to not work on non default streams? What about the most basic use for parallelizing a for-loop, i.e., launching a bunch of small kernels, one on each stream, and then letting the default stream wait on them?

**Comment by ngimel** [comment]
RNG works on non-default streams (as of a few weeks back, when mtgp was nuked and replaced by philox), thrust functions mostly do (I could have missed something in my grepping), magma does not https://github.com/pytorch/pytorch/issues/21821, and backward for anything does not.

**Comment by jfolz** [comment]
I believe I may have stumbled over this (a lot; over the course of a few weeks; and many scratched heads). Our code uses a background thread to move pinned tensors to GPU like so:
```
stream = cuda.Stream()
with cuda.stream(stream):
    while True:
        image, label = retrieve_data()
        image.to('cuda:0')
        label.to('cuda:0')
        cuda.synchronize()
        queue.put(tensors)
        queue.join()
```
`image` and `label` are a 128x3x224x224 uint8 tensor and 128 int64 tensor respectively.

Then in the foreground thread:
```
while True:
    tensors = queue.get()
    queue.task_done()
    output = model(tensors)
    ...
```

Where `...` is backward, optimizer step, etc. Under some very specific circumstances that I have yet to nail down, the GPU sees identical `image` tensors for batch 1 and 2 of every epoch. More precisely we are seeing the batch 2 tensor twice. `label` is not affected by this.
What I can say to far:
- The problem goes ways when forwarding within `no_grad`.
- I am not 100% sure that backward is necessary for it to appear.
- Everything works as expected when the stream is removed.

**Comment by jfolz** [comment]
The behavior in my previous comment happened with torchvision ResNet50 btw.
With a custom model we tried today it just outright crashed with `RuntimeError: cuda runtime error (700) : an illegal memory access was encountered` at random points within Pytorch.

**Comment by ezyang** [comment]
#22470 made me realize that it's not enough to audit CUDA kernel call sites for nullptr; if you don't pass enough arguments, that also counts as using the wrong stream too. We should write a lint rule for this.

**Comment by mruberry** [comment]
I enabled this flag on TestCUDA and TestNN in my branch with #8354 and the existing tests work fine. Ironically one of the tests I added fails (as expected) because it's doing something atypical, but I think there's a straightforward fix for that, too.

**Comment by mruberry** [comment]
See #25989, which runs the CI with the flag enabled in test_cuda.py and test_distributions.py. The failing tests appear unrelated.

**Comment by mruberry** [comment]
#25989 is landed and test_cuda.py and test_distrubtions.py now run on the non-default stream. Let's see how stable those tests are over the next few days.

**Comment by mruberry** [comment]
Status update: 

-cdist and repeat were not running on the current stream, fixed by @ngimel with two PRs: #26946, #26254.
-cuDNN RNNs are not syncing with the current stream, @ngimel intends to fix soon.
-ROCm tests were failing when run on a non-default stream even where non-ROCm CUDA tests were succeeding. All ROCm tests are currently run on the default stream.
-MAGMA tests were failing when run on a non-default stream. All MAGMA tests are currently run on the default stream. (MAGMA has an additional and severe issue, see #26996).

These issues are found as more tests are made generic, since generic CUDA tests are run on a non-default stream. The plan is to continue extending our generic test suites. Current coverage status:

- test_cuda, test_distrubtions, and test_indexing all run on the non-default stream
- most CUDA tests in test_torch.py run on the non-default stream, plan is to cover all CUDA tests in test_torch
- test_nn and test_autograd have a few generic tests, and the plan is to (again) continue extending coverage to all CUDA tests they define





 test_torch.py (mostl

**Comment by ezyang** [comment]
Deprioritizing as we've introduced non-default stream to many tests, and the rest are not high priority

**Comment by ngimel** [comment]
I suggest closing, we fixed the streaming bugs we were aware of and moved all the tests we wanted to run on a non-default stream to do so, additional bugs that can surface will need their own issues rather than this umbrella one.
